metric_learn/rca.py

```diff
@@ -62,32 +79,42 @@ def fit(self, data, chunks):
         when ``chunks[i] == j``, point i belongs to chunklet j.
     """
     data, chunks, num_chunks, d = self._process_inputs(data, chunks)
-
-    # mean center
-    data -= data.mean(axis=0)
-
-    # mean center each chunklet separately
-    chunk_mask = chunks != -1
-    chunk_data = data[chunk_mask]
-    chunk_labels = chunks[chunk_mask]
-    for c in xrange(num_chunks):
-      mask = chunk_labels == c
-      chunk_data[mask] -= chunk_data[mask].mean(axis=0)
-
-    # "inner" covariance of chunk deviations
+    chunk_mask, chunk_data, chunk_labels = _process_chunks(data, chunks, num_chunks)
     inner_cov = np.cov(chunk_data, rowvar=0, bias=1)
+    rank = np.linalg.matrix_rank(inner_cov)
+
+    # If the inner covariance matrix is not full rank,
+    # the input data are first projected with a PCA to a space of dimension rank.
+    M_pca = np.identity(d)
+    if rank < d:
+      pca = decomposition.PCA(n_components = rank)
+      pca.fit(data)
+      M_pca = pca.components_
+      data = pca.transform(data)
+      chunk_mask, chunk_data, chunk_labels = _process_chunks(data, chunks, num_chunks)
+      inner_cov = np.cov(chunk_data, rowvar=0, bias=1)
+      rank = np.linalg.matrix_rank(inner_cov)
```

Comment:
I like the idea to warn the user if the inner covariance matrix is not full rank, and to give advice about how to solve the issue.
However, if the user wants `d` dimensions, I don't think we should keep only the  `d` first principal components because we may lose some discriminative information. For instance, if the user wants only 2 dimensions for data visualization, keeping only the first two components is not enough : they may represent a very small part of the variance of the data, and the discriminative information may be in the unselected components.

My proposal :

1. Compute PCA and select the first components such that 90% of the variance is explained (easy to obtain with scikit-learn).
2. Compute the inner covariance matrix of the projected data and compute its rank using the standard non-null check.
3. If the inner covariance matrix is not full rank, then warn the user: 
"The inner covariance matrix is not invertible, and so the transformation matrix may contain Nan values. You should preprocess the data, with PCA for instance, to remove noise and redundant information."
We check that the inner covariance matrix is full rank, and not that `d <= r`. Indeed, if `d <= r` but the inner covariance matrix is not full rank, then we are going to select null eigenvalues (or very small ones).
4. If the inner covariance matrix is full rank, but `d > r`, warn the user that the embedding dimension has been reduced to `r`. 

What do you think of this proposal ?